docs/Customization/RoCC-Accelerators.rst

```diff
@@ -46,6 +50,47 @@ the ``busy`` signal, which indicates when the accelerator is still handling an i
 and the ``interrupt`` signal, which can be used to interrupt the CPU.
 
 Look at the examples in ``generators/rocket-chip/src/main/scala/tile/LazyRoCC.scala`` for detailed information on the different IOs.
+There is also more information about each of the signals in `the RoCC Documentation written by UCSD <https://docs.google.com/document/d/1CH2ep4YcL_ojsa3BVHEW-uwcKh1FlFTjH_kg5v8bxVw/edit>`_, although it is updated out of tree and may be out of date.
+
+
+Accessing Memory via L1 Cache
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+A RoCC accelerator can access memory through the L1 Cache of the core it is attached to.
+This is typically a good idea, as a RoCC instruction that passes in an address likely has accessed it recently.
+
+In your ``LazyRoCCModuleImp``, the signal ``io.mem`` is a ``HellaCacheIO``, which is defined in ``generators/rocket-chip/src/main/scala/rocket/HellaCache.scala``.
+If your core is a RocketCore, the other end of the input is connected to a ``SimpleHellaCacheIF``, which has a replay buffer in case the L1 cache is overloaded and doesn't ACK your request.
+
+.. code-block:: scala
+
+    class HellaCacheIO(implicit p: Parameters) extends CoreBundle()(p) {
+        val req = Decoupled(new HellaCacheReq)
+        val s1_kill = Output(Bool()) // kill previous cycle's req
+        val s1_data = Output(new HellaCacheWriteData()) // data for previous cycle's req
+        val s2_nack = Input(Bool()) // req from two cycles ago is rejected
+        val s2_nack_cause_raw = Input(Bool()) // reason for nack is store-load RAW hazard (performance hint)
+        val s2_kill = Output(Bool()) // kill req from two cycles ago
+        val s2_uncached = Input(Bool()) // advisory signal that the access is MMIO
+        val s2_paddr = Input(UInt(paddrBits.W)) // translated address
+
+        val resp = Flipped(Valid(new HellaCacheResp))
+        val replay_next = Input(Bool())
+        val s2_xcpt = Input(new HellaCacheExceptions)
+        val s2_gpa = Input(UInt(vaddrBitsExtended.W))
+        val s2_gpa_is_pte = Input(Bool())
+        val uncached_resp = tileParams.dcache.get.separateUncachedResp.option(Flipped(Decoupled(new HellaCacheResp)))
+        val ordered = Input(Bool())
+        val perf = Input(new HellaCachePerfEvents())
+
+        val keep_clock_enabled = Output(Bool()) // should D$ avoid clock-gating itself?
+        val clock_enabled = Input(Bool()) // is D$ currently being clocked?
+    }
+
+At a high level, you must tag requests that you send across this interface using the ``io.mem.req.tag``, and the tag will be returned to you when the data is ready.
+Responses may come back out of order if you issue multiple requests, so you can use these tags to tell what data came back.
+Note that the top two bits of the tag are reserved, and MUST be set to zero, or the interface will exhibit undefined behavior.
```

Comment:
Huh. I think that might actually be a bug somewhere.

Reply:
It's been a minute but from what I recall:

What happens is SimpleHellaCacheIF puts the entire tag into its replay queue, but the actual memory IF drops the top two bits... so when the data returns the tag is missing the top two bits and SimpleHellaCacheIF won't remove it from the nack queue (since it sees that it got a message with a different tag). This leads to the nack queue filling up and simplehellacacheif permanently asserting not ready.

If you look at the two rocc accelerators in chipyard they both set the top two bits to zero presumably for this reason. That's why I mentioned SimpleHellaCacheIF-- even if it should be an implementation detail you may need to inspect it to have working code 🥲

Reply:
What is the actual number of tag bits available to use? Is it 6?

Reply:
Looks like it's 

> (coreParams.dcacheReqTagBits + log2Ceil(dcacheArbPorts)

so maybe the top log2Ceil(dcacheArbPorts) bits of the tag are supposed to be the port that the request came in from and that's why they need to be zeroed(?)... I'll try to figure this out again 🤔

Reply:
I think it might just be dcachereqtagbits

Reply:
I updated this so we can get these docs in. Let me know if this still seems unclear.